std/algorithm/sorting.d

```diff
@@ -2626,6 +2626,173 @@ unittest
+        calls `medianOfMediansPartition` with half the range length.)
+
+Returns: `medianOfMediansPivot` returns an index `i` such that `r[i]` is within
+30th to 70th percentiles (i.e. in the middle 4 deciles). That means there are
+at least `0.3 * r.length` elements `x` in `r` that satisfy `!less(x, r[i])`
+and also at least `0.3 * r.length` elements `y` in `r` that satisfy
+`!less(r[i], y)`. `medianOfMediansPartition` returns `r[0 .. min($, nth)]`.
+*/
+size_t medianOfMediansPivot(alias less = "a < b", Range)(Range r)
+if (isRandomAccessRange!Range && hasLength!Range && hasSlicing!Range)
+{
+    alias lessFun = binaryFun!less;
+    if (r.length <= 5)
+    {
+        r.medianOf5OrFewer!lessFun;
+        return (r.length | 1) / 2;
+    }
+    // Take the median of each group of 5
+    size_t i = 0;
+    for (size_t j = 5; j < r.length; i = j, j += 5)
+    {
+        r[i .. j].medianOf5!lessFun;
+    }
+    // Leftovers
+    r[i .. $].medianOf5OrFewer!lessFun;
+    // Take a stride across all medians of five
+    import std.range : stride;
+    auto s = r[2 .. $].stride(5);
+    auto left = s.medianOfMediansPartition!lessFun(s.length / 2);
+    return 2 + 5 * left.length;
+}
+
+/// ditto
+Range medianOfMediansPartition(alias less = "a < b", Range)
+    (Range r, size_t nth)
+if (isRandomAccessRange!Range && hasLength!Range && hasSlicing!Range)
+{
+    if (nth >= r.length) return r[0 .. $];
+    auto result = r[0 .. nth];
+    for (;;)
+    {
+        auto pivot = r.medianOfMediansPivot!less;
+        assert(pivot < r.length);
+        assert(!binaryFun!less(r[pivot], r[pivot]));
+        import std.algorithm.mutation : swapAt;
+        r.swapAt(pivot, r.length - 1);
+        auto right = r.partition!(a => binaryFun!less(a, r.back));
+        assert(right.length >= 1);
+        pivot = r.length - right.length;
+        if (pivot > nth)
+        {
+            // We don't care to swap the pivot back, won't be visited anymore
+            assert(pivot < r.length);
+            r = r[0 .. pivot];
+            continue;
+        }
+        // Swap the pivot to where it should be
+        right.swapAt(0, right.length - 1);
+        if (pivot == nth)
+        {
+            // Found Waldo!
+            break;
+        }
+        ++pivot; // skip the pivot
+        r = r[pivot .. $];
+        nth -= pivot;
+    }
+    return result;
+}
+
+///
+unittest
+{
+    import std.algorithm : count, map, minPos;
+    import std.array : array;
+    import std.random : uniform;
+    import std.range : iota;
+    // Create an array of 100 random integers ranging between 0 and 1000.
+    auto v = iota(0, 100).map!(_ => uniform(0, 1_000)).array;
+    auto left = v.medianOfMediansPartition(v.length / 2);
+    // The largest element left of the median is less than or equal
+    assert(left.minPos!"a > b".front <= v[left.length]);
+    // The smallest element right of the median is greater than or equal
+    assert(v[left.length .. $].minPos.front >= v[left.length]);
+}
```

Comment:
The test are to easy, they are only testing the postive case. What about an empty range, what about a range with 3 elements. What about a range with custom types that support less.